src/eos-updater-poll-common.c

```diff
@@ -0,0 +1,910 @@
+  if (local_error != NULL &&
+      !g_error_matches (local_error,
+                        G_KEY_FILE_ERROR,
+                        G_KEY_FILE_ERROR_KEY_NOT_FOUND))
+    {
+      g_propagate_error (error, local_error);
+      return FALSE;
+    }
+
+  *out_on_hold = on_hold;
+  *out_ref = g_steal_pointer (&ref);
+  return TRUE;
+}
+
+static gboolean
+process_branch_file (GKeyFile *bkf,
+                     const gchar *group_name,
+                     gboolean *out_on_hold,
+                     gchar **out_ref,
+                     GError **error)
+{
+  /* Check for product-specific entry */
+  if (g_key_file_has_group (bkf, group_name))
+    {
+      message ("Product-specific branch configuration found");
+      if (!process_single_group (bkf, group_name, out_on_hold, out_ref, error))
+        return FALSE;
+      if (*out_on_hold)
+        message ("Product is on hold, nothing to upgrade here");
+      return TRUE;
+    }
+  /* Check for a DEFAULT_GROUP entry */
+  if (g_key_file_has_group (bkf, DEFAULT_GROUP))
+    {
+      message ("No product-specific branch configuration found, following %s",
+               DEFAULT_GROUP);
+      if (!process_single_group (bkf, DEFAULT_GROUP, out_on_hold, out_ref, error))
+        return FALSE;
+      if (*out_on_hold)
+        message ("No product-specific configuration and %s is on hold, "
+                 "nothing to upgrade here", DEFAULT_GROUP);
+      return TRUE;
+    }
+
+  *out_on_hold = FALSE;
+  *out_ref = NULL;
+  return TRUE;
+}
+
+gboolean
+get_upgrade_info_from_branch_file (EosBranchFile *branch_file,
+                                   gchar **upgrade_refspec,
+                                   gchar **original_refspec,
+                                   EosMetricsInfo **metrics,
+                                   GError **error)
+{
+  g_autoptr(OstreeDeployment) booted_deployment = NULL;
+  g_autofree gchar *booted_refspec = NULL;
+  g_autofree gchar *booted_remote = NULL;
+  g_autofree gchar *booted_ref = NULL;
+  g_autoptr(GHashTable) hw_descriptors = NULL;
+  g_autofree gchar *vendor = NULL;
+  g_autofree gchar *product = NULL;
+  g_autofree gchar *product_group = NULL;
+  g_autofree gchar *upgrade_ref = NULL;
+  gboolean on_hold = FALSE;
+
+  g_return_val_if_fail (EOS_IS_BRANCH_FILE (branch_file), FALSE);
+  g_return_val_if_fail (upgrade_refspec != NULL, FALSE);
+  g_return_val_if_fail (original_refspec != NULL, FALSE);
+
+  booted_deployment = eos_updater_get_booted_deployment (error);
+
+  if (!get_origin_refspec (booted_deployment, &booted_refspec, error))
+    return FALSE;
+
+  if (!ostree_parse_refspec (booted_refspec, &booted_remote, &booted_ref, error))
+    return FALSE;
+
+  hw_descriptors = get_hw_descriptors ();
+  vendor = cleanstr (g_strdup (g_hash_table_lookup (hw_descriptors, VENDOR_KEY)));
+  product = cleanstr (g_strdup (g_hash_table_lookup (hw_descriptors, PRODUCT_KEY)));
+
+  product_group = g_strdup_printf ("%s %s", vendor, product);
```

Comment:
It seems like `vendor` and `product` can be `NULL` here. What is the proper behaviour when they are `NULL`? It may be that `"(null) (null)"` is correct, but maybe there's a better option?